### ISSUE TITLE
Dump of positions gets stuck in Opensim

### ISSUE THREAD

**Issue by Anthonyghub** [AVsitter/AVsitter#45]
Hello, I have seen that the launch of the dump of positions in Opensim - Osgrid (in menu ADJUST - HELPER - DUMP) gets stuck after about 90 lines of saying the positions in local chat and there isn't the web dump of the web site neither. I have tried with different viewer, but it doesn't go forward all the times. Please help

**Comment by Sei-Lisa** [comment]
I can reproduce this. It looks like it's hitting a throttle or something alike.

**Comment by codeviolet** [comment]
OpenSim has a configurable scripted event timeout often set to 30 seconds that is probably the cause. I'm guessing that the script isn't being given time to loop through dumping all the poses.

I would think this can be overcome by changing ```EventLimit``` in ```OpenSim.ini```, but only the server owner can do so. My testing in Kitely didn't include large notecards. It's a shame to come up against such a heavy limitation which will probably require workarounds in every script that generates the dump data - something that may not be worth it just to support OpenSim.

**Comment by Anthonyghub** [comment]
Could it be possible to paste all the dump of AVpos in local chat all at once when you click the button Dump in one second? I have seen that older scripts like MLPV in opensim make the dump in one second and it works fine. Or could it possible to generate first the web dump link to avoid the timeout?

**Comment by codeviolet** [comment]
If the dump is timing out then the web link would also be incomplete. However, it may be a good idea to see if we can get it to go faster. OpenSim may be less prone to getting chat lines out of order than SL, so perhaps try removing the following line in the ```[AV]sitB``` script's readout section near the very bottom of the script: 
```lsl
llSleep(0.5);
```
Removing that line may hopefully enable dump to beat the event timeout clock, without any adverse consequences. Worth a try.

**Comment by Sei-Lisa** [comment]
I haven't had time to investigate this in more detail yet, but this seems like another reason in favour of implementing my suggestion in #14, which could benefit both the SL and the OpenSim versions.

@codeviolet: I added some quick debug code (I didn't have much time so I may have screwed up) and my impression was that it left the event and reentered it while dumping. I may be wrong, but if I'm right, then it can't be an event timeout, right?

**Comment by codeviolet** [comment]
@Sei-Lisa a sequence of events are triggered during the settings dump sequence, but it's possible that some of these stages are going over the limit. For example, when sitB receives ```90020``` during the settings dump sequence [here](https://github.com/AVsitter/AVsitter/blob/master/AVsitter2/%5BAV%5DsitB.lsl#L601) it iterates though all the poses for that sitter in a ```link_message``` event, with a llSleep(0.5) for each pose and llSleep(0.2) for each position/rotation. So if there were 50 poses for the sitter, that's already probably going to go over the limit. I agree perhaps it can go hand in hand with #14.

The two llSleep() lines at the bottom of sitB are only to prevent the chat lines getting of order in SL. I don't know if they have are needed in OpenSim. That's why I thought it worth trying to remove them, to see if the script then beats the event timeout.

**Comment by Anthonyghub** [comment]
I have heard that llSleep doesn't work fine in Opensim and it can cause problems; and they recommend to use llSetTimerEvent ; but i don't know really

**Comment by Sei-Lisa** [comment]
Thanks, I edited my message when I realized my mistake. You were completely right, I was just looking at the wrong script.

This patch fixes the issue for me:

```diff
--- [AV]sitB.oss	2017-09-18 00:27:13.161462929 +0200
+++ [AV]sitB.oss.patched	2017-09-18 17:26:55.957723386 +0200
@@ -63,11 +63,6 @@
     llMessageLinked(LINK_THIS, 90055, (string)SCRIPT_CHANNEL, llDumpList2String([llList2String(MENU_LIST, ANIM_INDEX), llList2String(DATA_LIST, ANIM_INDEX), llList2String(POS_ROT_LIST, ANIM_INDEX * 2), llList2String(POS_ROT_LIST, ANIM_INDEX * 2 + 1), broadcast, speed_index], "|"));
 }
 
-Readout_Say(string say)
-{
-    llMessageLinked(LINK_THIS, 90022, say, (string)SCRIPT_CHANNEL);
-}
-
 list order_buttons(list buttons)
 {
     return llList2List(buttons, -3, -1) + llList2List(buttons, -6, -4) + llList2List(buttons, -9, -7) + llList2List(buttons, -12, -10);
@@ -600,23 +595,35 @@
             }
             else if (num == 90020 && llList2String(data, 0) == "")
             {
-                Readout_Say("V:" + llDumpList2String([version, MTYPE, ETYPE, SET, SWAP, SITTER_INFO, CUSTOM_TEXT, ADJUST_MENU, SELECT, AMENU, OLD_HELPER_METHOD], "|"));
-                integer i = -1;
-                while (++i < llGetListLength(MENU_LIST))
+                llMessageLinked(LINK_THIS, 90022, "V:" + llDumpList2String([version, MTYPE, ETYPE, SET, SWAP, SITTER_INFO, CUSTOM_TEXT, ADJUST_MENU, SELECT, AMENU, OLD_HELPER_METHOD], "|"), (string)SCRIPT_CHANNEL);
+                llMessageLinked(LINK_THIS, 90024, (string)SCRIPT_CHANNEL, "-1|D");
+            }
+            else if (num == 90024) // self-sent message to dump the next line
+            {
+                if (llList2String(data, 1) == "D")
+                {
+                if (++two < llGetListLength(MENU_LIST))
                 {
                     llSleep(0.5);
-                    Readout_Say("S:" + llList2String(MENU_LIST, i) + "|" + llList2String(DATA_LIST, i));
+                    llMessageLinked(LINK_THIS, 90022, "S:" + llList2String(MENU_LIST, two) + "|" + llList2String(DATA_LIST, two), (string)SCRIPT_CHANNEL);
+                    llMessageLinked(LINK_THIS, 90024, (string)SCRIPT_CHANNEL, (string)two + "|D");
+                    return;
+                }
+                llMessageLinked(LINK_THIS, 90024, (string)SCRIPT_CHANNEL, "-1|P");
+                return;
                 }
-                i = -1;
-                while (++i < llGetListLength(MENU_LIST))
+                if (++two < llGetListLength(MENU_LIST))
                 {
-                    if (llList2Vector(POS_ROT_LIST, i * 2) != ZERO_VECTOR)
+                    if (llList2Vector(POS_ROT_LIST, two * 2) != ZERO_VECTOR)
                     {
                         llSleep(0.2);
-                        Readout_Say("{" + llList2String(MENU_LIST, i) + "}" + llList2String(POS_ROT_LIST, i * 2) + llList2String(POS_ROT_LIST, i * 2 + 1));
+                        llMessageLinked(LINK_THIS, 90022, "{" + llList2String(MENU_LIST, two) + "}" + llList2String(POS_ROT_LIST, two * 2) + llList2String(POS_ROT_LIST, two * 2 + 1), (string)SCRIPT_CHANNEL);
                     }
+                    llMessageLinked(LINK_THIS, 90024, (string)SCRIPT_CHANNEL, (string)two + "|P");
+                    return;
                 }
                 llMessageLinked(LINK_THIS, 90021, (string)SCRIPT_CHANNEL, "");
+                return;
             }
         }
     }
```

(the indentation isn't 100% right, because I've kept the original indentation to help see the changes). The idea is straightforward: loop via link messages rather than with a `while`.

My idea to implement #14 as extra scripts was similar to this, but moving the (new) 90024 messages to the new dump script; this way, the dump script can take control of the timing and it can even afford to report progress in hovering test. Concerning memory, my prediction is that the removal of Readout_Say will probably compensate the code added, and after the change in #14, which implies the removal of two or three extra llMessageLinked, it will probably be a net gain.

I've had a chance to meet Anthonyghub inworld and passed him the [AV]stB with the above patch so he can perform dumps. He confirmed it worked for him too.

Thanks @Anthonyghub for the report!

**Comment by Anthonyghub** [comment]
Great job! Thank you very much! :)

**Comment by codeviolet** [comment]
@sei-lisa I recall having some concern about creating a system where if only 1 of "many" link messages were somehow missed that the dump would halt... but if LSL occasionally dropping link messages is not a problem (it may only ever be something I had imagined) then I am really impressed by this solution.

Side notes:

* It will be good to get a patched OpenSim "release" out when convenient to avoid other people having this problem with settings dump.

* It may be logical to apply this same fix to [AV]faces & [AV]prop before OpenSim release, which also iterate through each item in their memory during the dump sequence in a similar way.

**Comment by Sei-Lisa** [comment]
Regarding release plans:

My idea is to ship a modified version of the above patch to make it OpenSim-only, as a quick fix for OS users to have something working, and to give us time to work on doing things right. I'm quite busy at this time and can't take care of doing the full fix yet, but I expect I can work on it in about a week.

My idea of "doing things right" is to implement #14 for all scripts. That includes [AV]adjuster, [AV]faces, [AV]prop, [AV]camera, [AV]menu, AVpos-generator, AVpos-shifter and MLP-converter, if I'm not mistaken. I guess Noob-detector should be left out because of its nature.

Re link messages:

I've made reliability tests with link messages, one for possible loss of messages and another for possible out-of-order reception. They were in an attachment that I traveled with for days, and there wasn't a single failure in either category. The only potential cause for loss of messages is filling the event queue, which has a capacity for 64 messages. I think it's pretty safe. Anyway, even if one were lost in this case, the user would notice and can restart the dump.
